python/ml4ir/base/model/calibrations/temperature_scaling.py

```diff
@@ -0,0 +1,247 @@
+        labels_tensor = tf.constant(labels_nps, name='labels_tensor', dtype=tf.int32)
+        logits_w_temp = tf.divide(logits_tensor, temperature)
+
+        return tf.reduce_mean(
+            tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels_tensor,
+                                                           logits=logits_w_temp))
+
+    @tf.function
+    def nll_with_lbfgs():
+        return tfp.optimizer.lbfgs_minimize(make_val_and_grad_fn(nll_loss_fn),
+                                            initial_position=start)
+
+    # evaluation on validation set before temperature scaling
+    logits_nps, labels_nps = get_logits_labels(intermediate_model, dataset.validation)
+    org_acc_op, org_nll_loss_op, _ = eval_relevance_model(scorer, logits_nps,
+                                                          labels_nps)
+
+    # perform temperature scaling
+    results = run(nll_with_lbfgs, logger)
+
+    # evaluation on validation set with temperature scaling
+    temper = tf.constant(results.position)
+    acc_op, nll_loss_op, _ = eval_relevance_model(scorer, logits_nps, labels_nps,
+                                                  temperature=temper)
+
+    logger.info("=" * 50)
+    logger.info(f'temperature value : {results.position}')
+    logger.info('Evaluating on the validation dataset ')
+    logger.info(f'original loss: {org_nll_loss_op}, accuracy: {org_acc_op}, \n'
+          f'temperature scaling loss: {results.objective_value}, accuracy: {acc_op}\n')
+
+    logger.info("="*50)
+    logger.info("Evaluating on the test dataset")
+
+    logits_nps_test, labels_nps_test = get_logits_labels(intermediate_model, dataset.test)
+
+    # evaluation on test set before temperature scaling
+    acc_test_ts, _, org_softmaxes = eval_relevance_model(scorer, logits_nps_test,
+                                          labels_nps_test)
+
+    # evaluation on test set with temperature scaling
+    acc_test_org, _, ts_softmaxes = eval_relevance_model(scorer,
+                                                         logits_nps_test,
+                                           labels_nps_test, temperature=temper)
+
+    # write the full vector in the csv not ...
+    np.set_printoptions(
+        formatter={'all': lambda x: str(x.decode('utf-8')) if isinstance(x, bytes) else str(x)},
+        linewidth=sys.maxsize,
+        threshold=sys.maxsize)
+    # avoiding .tolist() that is not memory efficient
+    data_dic = {'org_scores': [x for x in org_softmaxes.numpy()],
+                'ts_scores': [x for x in ts_softmaxes.numpy()],
+                'org_predicted_label': tf.argmax(org_softmaxes, axis=-1).numpy(),
+                'true_label': labels_nps_test,
+                }
+
+    logger.info(f'original test accuracy: {acc_test_org}, \ntemperature scaling '
+                     f'test accuracy: {acc_test_ts} \n')
+
+    # the file is big and should be zipped
+    zip_dir_path = dict_to_zipped_csv(data_dic, logs_dir_local)
```

Comment:
@marjanhs I don't understand why we need to save this file here. Can you link a reference for me? Sorry :)

Reply:
@lastmansleeping  we would like to know if the output probabilities of the test set  are well calibrated compared to the original probabilities (before calibration). So, we are saving them to plot reliability diagrams and compute average errors  (ECE) and maximum Error (MCE) later.

Reply:
Okay. That makes sense. Thanks!

Do you plan to add the plots and metrics to ml4ir as well? I think it might be useful? Otherwise, I'm assuming the current workflow involves training a model and then running a post-eval script? That should be fine too, as long as you check it in somewhere. ;)

Reply:
I recommend doing this in a follow up PR. Can be in one of our internal repos as well if you think it contains salesforce-specific info.

Reply:
@balikasg thoughts?

Reply:
@lastmansleeping  we can add them to ml4ir, but I don't think it is a priority as the code(in pytorch) is ready and can be used independently of `Tensorflow`. It only needs the `.csv` file we are generating here. I should add the code to an internal repo for now (and if gets priority convert it to Tensorflow  and add it here).

Reply:
post-eval in pytorch? What does this post-eval script do?

Reply:
@jakemannix  it computes the  inputs of reliability diagram, ECE and MCE .